src/python/pants/backend/javascript/subsystems/nodejs.py

```diff
@@ -201,18 +245,83 @@ class NodeJsBootstrap:
     nodejs_search_paths: tuple[str, ...]
 
 
+async def _nodejs_search_paths(env_tgt: EnvironmentTarget, paths: Collection[str]) -> tuple[str, ...]:
+    asdf_result = await AsdfToolPathsResult.get_un_cachable_search_paths(
+        paths,
+        env_tgt=env_tgt,
+        tool_name="nodejs",
+        tool_description="Node.js distribution",
+        paths_option_name="[nodejs].search_path",
+    )
+    asdf_standard_tool_paths = asdf_result.standard_tool_paths
+    asdf_local_tool_paths = asdf_result.local_tool_paths
+    special_strings: dict[str, Iterable[str]] = {
+        AsdfPathString.STANDARD: asdf_standard_tool_paths,
+        AsdfPathString.LOCAL: asdf_local_tool_paths,
+    }
+
+    expanded: list[str] = []
+    for s in paths:
+        if s == "<PATH>":
+            expanded.extend(await Get(PathEnvironmentVariable, {}))  # noqa: PNT30: Linear search
+        elif s in special_strings:
+            expanded.extend(special_strings[s])
+        else:
+            expanded.append(s)
+    return tuple(expanded)
+
+
 @rule
-async def nodejs_bootstrap() -> NodeJsBootstrap:
-    return NodeJsBootstrap(nodejs_search_paths=())
+async def nodejs_bootstrap(nodejs_env_aware: NodeJS.EnvironmentAware) -> NodeJsBootstrap:
+    search_paths = await Get(
+        ValidatedSearchPaths,
+        ValidateSearchPathsRequest(
+            env_tgt=nodejs_env_aware.env_tgt,
+            search_paths=tuple(nodejs_env_aware.search_path),
+            option_origin=f"[{NodeJS.options_scope}].search_path",
+            environment_key="nodejs_search_path",
+            is_default=nodejs_env_aware._is_default("search_path"),
+            local_only=FrozenOrderedSet((AsdfPathString.STANDARD, AsdfPathString.LOCAL)),
+        ),
+    )
+
+    expanded_paths = await _nodejs_search_paths(nodejs_env_aware.env_tgt, search_paths)
+
+    return NodeJsBootstrap(nodejs_search_paths=expanded_paths)
+
+
+class _BinaryPathsPerVersion(FrozenDict[str, tuple[BinaryPath, ...]]):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Testing for Node.js binaries.")
+async def get_valid_nodejs_paths_by_version(bootstrap: NodeJsBootstrap) -> _BinaryPathsPerVersion:
+    paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(
+            search_path=bootstrap.nodejs_search_paths,
+            binary_name="node",
+            test=BinaryPathTest(
+                ["--version"], fingerprint_stdout=False
+            ),  # Hack to retain version info
```

Comment:
Is this too hacky?

Reply:
More the name of the variable that gave me pause, I think. It's called "fingerprint" but is the `--version` output.